### ISSUE TITLE
question on the cross validation error logging in the training after applying useEvalCriterionControlLR=true

### ISSUE THREAD

**Issue by funguy74** [Microsoft/CNTK#635]
Since I closed prior issue too soon, I'll continue this with new thread.

Briefly reminding what my issue is that

I want to display cross validation error in the training log. Usually, two following ports print out two cost functions in the ndl file.

cr = CrossEntropyWithSoftmax(labels, LSTMoutputW,tag=Criteria);
Err = ErrorPrediction(labels,LSTMoutputW,tag=Eval);

What I like to do is to let ErrorPrediction use cross validation data for error calculation. From prior suggestion, if I set useEvalCriterionControlLR to be true, ErrorPrediction will use cross validation data (set by cvReader).

However, even after I applied this, still ErrorPrediction use training set. I can confirm this by following log:

> Epoch[ 1 of 50]-Minibatch[30901-31000]: cr = 3.09460000 * 80000; Err = 0.62770000 * 80000; time = 22.3754s; samplesPerSecond = 3575.4
> releaserandomizedchunk: paging out randomized chunk 270 (frame range [24303867..24393906]), 6 resident in RAM
>  Epoch[ 1 of 50]-Minibatch[31001-31100]: cr = 3.19330000 * 80000; Err = 0.63480000 * 80000; time = 22.3220s; samplesPerSecond = 3583.9
>  Epoch[ 1 of 50]-Minibatch[31101-31200]: cr = 3.12385714 * 56000; Err = 0.63450000 * 56000; time = 22.6078s; samplesPerSecond = 2477.0
> Finished Epoch[ 1 of 50]: [Training] **cr = 3.79490976 * 24937023**; **Err = 0.69966996 * 24937023**; totalSamplesSeen = 24937023; learningRatePerSample = 0.00012500001; epochTime=11115.2s

If you look at cr and err highlighted, both use the same size data (24937023). So, I think ErrorPrediction use same training data. 

Any idea what's going on? I added my config file in case you are curious (it is basically ami config)

```
NdlDir=$ExpDir$/configs
command=$command$
numCPUThreads=$numThreads$
precision=float

Truncated=true
TrainModel=[
    action=trainRNN
    modelPath=$ExpDir$/$modelName$

    # deviceId=-1 for CPU, >=0 for GPU devices
    deviceId=$DeviceNumber$
    traceLevel=1
    useValidation=true
    NDLNetworkBuilder=[
        ndlMacros=$NdlDir$/default_macros.ndl
        networkDescription=$NdlDir$/model.ndl
    ]


    SGD=[
        epochSize=0
        minibatchSize=$minibatchSize$
        learningRatesPerMB=$lratePerMB$
        momentumPerMB=$momentum$
        L2RegWeight=$l2RegWeight$
        dropoutRate=$dropoutRate$
        maxEpochs=$maxEpochs$
        numMBsToShowResult=$evaluatePeriod$


        #settings for Auto Adjust Learning Rate
        AutoAdjust=[
            reduceLearnRateIfImproveLessThan=0
            loadBestModel=true
            increaseLearnRateIfImproveMoreThan=1000000000
            learnRateDecreaseFactor=0.5
            autoAdjustLR=AdjustAfterEpoch
            learnRateAdjustInterval=1
            useEvalCriterionControlLR=true
        ]

        gradientClippingWithTruncation=true
        clippingThresholdPerSample=$clippingThresholdPerSample$

    ]
    # Parameter values for the reader
    reader=[
        # reader to use
        readerType=Kaldi2Reader
        readMethod=blockRandomize
        frameMode=false
        Truncated=true
        nbruttsineachrecurrentiter=$numUttsPerMinibatch$
        miniBatchMode=Partial
        randomize=1000000
        verbosity=0
        features=[
            dim=$featDim$
            scpFile=$inputCounts$
            rx=$inputFeats$
            featureTransform=$featureTransform$
        ]

        labels=[
            mlfFile=$inputLabels$
            labelDim=$labelDim$
            labelMappingFile=$labelMapping$
        ]
    ]
    cvReader=[
        # reader to use
        readerType=Kaldi2Reader
        readMethod=blockRandomize
        frameMode=false
        miniBatchMode=Partial
        randomize=Auto
        Truncated=true
        nbruttsineachrecurrentiter=64

        verbosity=0
        features=[
            dim=$featDim$
            scpFile=$cvInputCounts$
            rx=$cvInputFeats$
            featureTransform=$featureTransform$
        ]

        labels=[
            mlfFile=$cvInputLabels$
            labelDim=$labelDim$
            labelMappingFile=$labelMapping$
        ]
    ]
]
```

**Comment by dongyu888** [comment]
After each training epoch, the SGD will evaluate the current model on the CV set. Don’t you see something like

 

Finished Epoch[ 1 of 50]: [Validate]

 

Before the next epoch?

**Comment by funguy74** [comment]
The following is what you are saying.
> Final Results: Minibatch[1-1980]: cr = 3.39574755 * 2489538; perplexity = 29.83694989; Err = 0.67587641 * 2489538
> Finished Epoch[ 1 of 50]: [Validate] cr = 3.39574755 * 2489538; Err = 0.67587641 * 2489538

Still, both show same number of samples.

**Comment by dongyu888** [comment]
Your training set is 
Finished Epoch[ 1 of 50]: [Training] cr = 3.79490976 * 24937023; Err = 0.69966996 * 24937023; totalSamplesSeen = 24937023; learningRatePerSample = 0.00012500001; epochTime=11115.2s

cv set is
Finished Epoch[ 1 of 50]: [Validate] cr = 3.39574755 * 2489538; Err = 0.67587641 * 2489538

why do you say the sample size is the same?

**Comment by funguy74** [comment]
I understand what you mean. Initially, I thought cr is calculated based on training set and err will be calculated based on validation set. So, my expectation is to compare validation and training curve simultaneously during training stage.
However, what you are saying is that both cr and err are calculated based on training set most of time during epoch and only at the end of each epoch, cr and err will be calculated based on CV. Right?

If then, is there any way we can display validation error not only at the end of epoch but also during training? Because, I think one at each epoch looks a bit coarse and it would be better we can control this interval.

**Comment by dongyu888** [comment]
No. CV has different size than training set and each evaluation on CV slows down the training process. If you want to evaluate CV more just reduce the epoch size.

**Comment by funguy74** [comment]
Thanks for your answer. 

However, if we reduce epoch size, isn't it the same as reducing training data?  Is there any way we can change epoch size but using full training data in CNTK?

**Comment by dongyu888** [comment]
No, just each sweep of data take more epochs.

**Comment by funguy74** [comment]
Great! Thanks for clarification. I think one solution you suggest is viable by reducing number of sample sweep per each epoch.

Thanks a lot
